pkg/bridgeclient/app.go

```diff
@@ -224,69 +206,115 @@ func getMessagesFromCloud(serverURL, clientID string) ([]v1.BridgeMessage, error
 
-func sendMessageToCloud(msg *nats.Msg, serverURL string, clientID string, ceEnabled bool) {
-	log.Debugf("Sending Msg NB %s", msg.Subject)
-
-	msgFormat := msgs.GetMsgFormat()
-	status, err := msgFormat.ValidateMsgFormat(msg.Data, ceEnabled)
+func subscribeToOutboundMessages(serverURL, clientID string) (*nats.Subscription, error) {
+	nc := bridgemodel.GetNatsConnection()
+	subj := fmt.Sprintf("%s.>", msgs.NATSSYNC_MESSAGE_PREFIX)
+	sub, err := nc.SubscribeSync(subj)
 	if err != nil {
-		log.Errorf("Error validating the cloud event message: %s", err.Error())
-		return
+		return nil, err
+	}
+	go handleOutboundMessages(sub, serverURL, clientID)
+	return sub, nil
+}
+
+// handleOutboundMessages  This pulls messages off the queue and groups a bunch of them to push them together
+// if we have to wait more than N ms for a message, we will go ahead and send what we have
+// or if we get more than N messages, we will send them along
+func handleOutboundMessages(subscription *nats.Subscription, serverURL, clientID string) {
+	timeoutStr := pkg.GetEnvWithDefaults("NATSSYNC_MSG_WAIT_TIMEOUT", "5")
+	maxMsgHoldStr := pkg.GetEnvWithDefaults("NATSSYNC__MAX_MSG_HOLD", "512")
+	waitTimeout, numErr := strconv.ParseInt(timeoutStr, 10, 16)
+	if numErr != nil {
+		waitTimeout = 5
 	}
-	if !status {
-		log.Errorf("Cloud event message validation failed, ignoring the message...")
-		return
+	maxQueueSize, numErr := strconv.ParseInt(maxMsgHoldStr, 10, 16)
+	if numErr != nil {
+		waitTimeout = 512
 	}
 
-	url := fmt.Sprintf("%s/bridge-server/1/message-queue/%s", serverURL, clientID)
-	natmsg := bridgemodel.NatsMessage{Reply: msg.Reply, Subject: msg.Subject, Data: msg.Data}
-	envelope, enverr := msgs.PutObjectInEnvelope(natmsg, clientID, pkg.CLOUD_ID)
-	if enverr != nil {
-		log.Errorf("Error putting msg in envelope %s", enverr.Error())
-		return
+	msgList := make([]*nats.Msg, 0)
+	keepGoing := true
+	for keepGoing {
+		msg, err := subscription.NextMsg(time.Duration(waitTimeout) * time.Millisecond)
+		sendWhatWeHave := false
+		//if we get a timeout (or any error) send what we have
+		if err != nil {
+			sendWhatWeHave = len(msgList) > 0
+		} else {
+			parsedSubject, err2 := msgs.ParseSubject(msg.Subject)
+			if err2 == nil {
+				log.Tracef("Found message to send NB Stored  Client ID=%s, Message Target %s", clientID, parsedSubject.LocationID)
+				//if the target client ID is not this client, push it to the server
+				if parsedSubject.LocationID != clientID {
+					log.Tracef("Adding message to list to send NB")
+					msgList = append(msgList, msg)
+				} else {
+					log.Tracef("Message not meant for NB, dropping")
+				}
+			}
+			sendWhatWeHave = len(msgList) > int(maxQueueSize)
+		}
+		if sendWhatWeHave {
+			sendMessageToCloud(serverURL, clientID, false, msgList...)
+			msgList = make([]*nats.Msg, 0)
+		}
 	}
+	log.Infof("Leaving Handle Outbound Messages ")
+}
+func sendMessageToCloud(serverURL string, clientID string, ceEnabled bool, msgsList ...*nats.Msg) {
```

Comment:
Just curious, why is `msgsList` defined as `...*nats.Msg` instead of `[]*nats.Msg`?

Reply:
calling convience.  you can pass nats messages as non arrays if you have only 1 of them (or two)